call_test.go

```diff
@@ -213,6 +215,208 @@ func setUp(t *testing.T, port int, maxStreams uint32) (*server, *ClientConn) {
 	return server, cc
 }
 
+func setUpWithOptions(t *testing.T, port int, maxStreams uint32, dopts ...DialOption) (*server, *ClientConn) {
+	server := newTestServer()
+	go server.start(t, port, maxStreams)
+	server.wait(t, 2*time.Second)
+	addr := "localhost:" + server.port
+	dopts = append(dopts, WithBlock(), WithInsecure(), WithCodec(testCodec{}))
+	cc, err := Dial(addr, dopts...)
+	if err != nil {
+		t.Fatalf("Failed to create ClientConn: %v", err)
+	}
+	return server, cc
+}
+
+func (s) TestUnaryClientInterceptor(t *testing.T) {
+	parentKey := ctxKey("parentKey")
+
+	interceptor := func(ctx context.Context, method string, req, reply interface{}, cc *ClientConn, invoker UnaryInvoker, opts ...CallOption) error {
+		if ctx.Value(parentKey) == nil {
+			t.Fatalf("interceptor should have %v in context", parentKey)
+		}
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+
+	server, cc := setUpWithOptions(t, 0, math.MaxUint32, WithUnaryInterceptor(interceptor))
+	defer func() {
+		cc.Close()
+		server.stop()
+	}()
+
+	var reply string
+	ctx := context.Background()
+	parentCtx := context.WithValue(ctx, ctxKey("parentKey"), 0)
+	if err := cc.Invoke(parentCtx, "/foo/bar", &expectedRequest, &reply); err != nil || reply != expectedResponse {
+		t.Fatalf("grpc.Invoke(_, _, _, _, _) = %v, want <nil>", err)
+	}
+}
+
+func (s) TestChainUnaryClientInterceptor(t *testing.T) {
+	var (
+		parentKey    = ctxKey("parentKey")
+		firstIntKey  = ctxKey("firstIntKey")
+		secondIntKey = ctxKey("secondIntKey")
+	)
+
+	firstInt := func(ctx context.Context, method string, req, reply interface{}, cc *ClientConn, invoker UnaryInvoker, opts ...CallOption) error {
+		if ctx.Value(parentKey) == nil {
+			t.Fatalf("first interceptor should have %v in context", parentKey)
+		}
+		if ctx.Value(firstIntKey) != nil {
+			t.Fatalf("first interceptor should not have %v in context", firstIntKey)
+		}
+		if ctx.Value(secondIntKey) != nil {
+			t.Fatalf("first interceptor should not have %v in context", secondIntKey)
+		}
+		firstCtx := context.WithValue(ctx, firstIntKey, 1)
+		return invoker(firstCtx, method, req, reply, cc, opts...)
+	}
+
+	secondInt := func(ctx context.Context, method string, req, reply interface{}, cc *ClientConn, invoker UnaryInvoker, opts ...CallOption) error {
+		if ctx.Value(parentKey) == nil {
+			t.Fatalf("second interceptor should have %v in context", parentKey)
+		}
+		if ctx.Value(firstIntKey) == nil {
+			t.Fatalf("second interceptor should have %v in context", firstIntKey)
+		}
+		if ctx.Value(secondIntKey) != nil {
+			t.Fatalf("second interceptor should not have %v in context", secondIntKey)
+		}
+		secondCtx := context.WithValue(ctx, secondIntKey, 2)
+		return invoker(secondCtx, method, req, reply, cc, opts...)
+	}
+
+	lastInt := func(ctx context.Context, method string, req, reply interface{}, cc *ClientConn, invoker UnaryInvoker, opts ...CallOption) error {
```

Comment:
Being a bit paranoid:
We should also make sure this interceptor is called. One way is to append something to the end of reply, and then check that `reply == expectedResponse + surfix`.

Reply:
done